### PR TITLE
executor, plan: support batch index look up join.

### PR DESCRIPTION
**Pull request by hanfei1991**
Implement the batch index look up join and the batch size is controlled by index look up size. Some performance bench result will be done in the future.

@shenli @zimulala @winoros @XuHuaiyu PTAL

### CODE REVIEW

**Comment by shenli** [comment]
@hanfei1991 Please fix CI.

File: executor/new_distsql.go [hunk 1]
```diff
@@ -266,6 +295,20 @@ func (e *IndexLookUpExecutor) Open() error {
 	return nil
 }
 
+// doRequestForHandles constructs kv ranges by handles. It is used by index look up executor.
```
Comment (XuHuaiyu) [inline executor/new_distsql.go]: s/doRequestForHandles/doRequestForDatums

File: executor/new_distsql.go [hunk 2]
```diff
@@ -217,6 +232,20 @@ func (e *IndexReaderExecutor) Open() error {
...
+func (e *IndexReaderExecutor) doRequestForHandles(values [][]types.Datum, goCtx goctx.Context) error {
```
Comment (XuHuaiyu) [inline executor/new_distsql.go]: s/doRequestForHandles/doRequestForDatums ?

File: executor/distsql.go [hunk 3]
```diff
@@ -156,6 +156,20 @@ func tableHandlesToKVRanges(tid int64, handles []int64) []kv.KeyRange {
 	return krs
 }
 
+func indexValuesToKVRanges(tid, idxID int64, values [][]types.Datum) ([]kv.KeyRange, error) {
+	krs := make([]kv.KeyRange, 0, len(values))
+	for _, vals := range values {
+		// TODO: We don't process the case that equal key has different types.
+		valKey, err := codec.EncodeKey(nil, vals...)
+		if err != nil {
+			return nil, errors.Trace(nil)
```
Comment (XuHuaiyu) [inline executor/distsql.go]: errors.Trace(err)？

File: executor/distsql.go [hunk 4]
```diff
@@ -156,6 +156,20 @@ func tableHandlesToKVRanges(tid int64, handles []int64) []kv.KeyRange {
 	return krs
 }
 
+func indexValuesToKVRanges(tid, idxID int64, values [][]types.Datum) ([]kv.KeyRange, error) {
```
Comment (XuHuaiyu) [inline executor/distsql.go]: Comment for this function.

File: executor/new_join.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+		e.innerRows = append(e.innerRows, orderRow{key: joinKey, row: innerRow})
+	}
+	sort.Sort(e.innerRows)
+	return e.doMergeJoin()
+}
+
+func getNextCursor(cursor int, rows orderedRows) int {
```
Comment (XuHuaiyu) [inline executor/new_join.go]: comment for this function.

File: executor/new_distsql.go [hunk 6]
```diff
@@ -127,6 +134,15 @@ func (e *TableReaderExecutor) doRequestForHandles(handles []int64, goCtx goctx.C
 	return nil
 }
 
+// doRequestForDatums constructs kv ranges by Datums. It is used by index look up executor.
+func (e *TableReaderExecutor) doRequestForDatums(datums [][]types.Datum, goCtx goctx.Context) error {
```
Comment (XuHuaiyu) [inline executor/new_distsql.go]: comment for why choose datum[0]

File: plan/new_physical_plan_builder.go [hunk 7]
```diff
@@ -148,7 +148,6 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
 func (p *LogicalJoin) convertToIndexJoin(prop *requiredProp, outerIdx int) (taskProfile, error) {
 	outerChild := p.children[outerIdx].(LogicalPlan)
 	innerChild := p.children[1-outerIdx].(LogicalPlan)
-	canPassProp := len(outerChild.Schema().ColumnsIndices(prop.cols)) > 0
```
Comment (XuHuaiyu) [inline plan/new_physical_plan_builder.go]: why not check this?

File: executor/new_join.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+	"bytes"
+	"sort"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type orderRow struct {
```
Comment (shenli) [inline executor/new_join.go]: orderRow -> orderedRow

File: executor/new_join.go [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+	e.exhausted = false
+	return errors.Trace(e.children[0].Open())
+}
+
+// Close implements the Executor Close interface.
+func (e *IndexLookUpJoin) Close() error {
+	e.resultRows = nil
+	e.outerRows = nil
+	e.innerDatums = nil
+	return errors.Trace(e.children[0].Close())
```
Comment (shenli) [inline executor/new_join.go]: Need to close innerExec?

File: executor/new_join.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+// Next implements the Executor Next interface.
+func (e *IndexLookUpJoin) Next() (*Row, error) {
```
Comment (shenli) [inline executor/new_join.go]: Add a few comment in this function.

File: executor/new_join.go [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+		if i > 0 && bytes.Compare(rows[i-1].key, rows[i].key) == 0 {
+			continue
+		}
+		datums = append(datums, rows[i].row.Data)
+	}
+	return datums
+}
+
+func (e *IndexLookUpJoin) doJoin() error {
+	err := e.innerExec.doRequestForDatums(uniqueOrderedRows(e.innerDatums), e.ctx.GoCtx())
```
Comment (shenli) [inline executor/new_join.go]: Should we call e.innerExec.Open()?

File: executor/new_join.go [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+		}
+		c := bytes.Compare(rows[cursor].key, rows[cursor-1].key)
+		if c != 0 {
+			break
+		}
+	}
+	return cursor
+}
+
+func (e *IndexLookUpJoin) doMergeJoin() error {
```
Comment (shenli) [inline executor/new_join.go]: Add a few comments for this function. Such as:
e.innerRows and e.outerRows are already sorted, so we can do SortMergeJoin on the two data source.

File: executor/new_join.go [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+				e.resultRows = append(e.resultRows, e.fillDefaultValues(outerRow))
+			}
+		} else {
+			innerCursor = getNextCursor(innerCursor, e.outerRows)
+		}
+	}
+	for e.outer && outerCursor < len(e.outerRows) {
+		outerRow := e.outerRows[outerCursor].row
+		e.resultRows = append(e.resultRows, e.fillDefaultValues(outerRow))
+		outerCursor++
```
Comment (shenli) [inline executor/new_join.go]: outerCursor = getNextCursor(outerCursor, e.outerRows)?

File: plan/new_physical_plan_builder.go (hunk 7 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: The batch index join cannot inherit the properties of orders.

File: executor/new_join.go (hunk 9 above)
Comment (hanfei1991) [inline executor/new_join.go]: innerExec has been opened and closed for multiple times.

File: executor/new_join.go (hunk 11 above)
Comment (hanfei1991) [inline executor/new_join.go]: Everytime we execute an Executor, we should call Open().

File: executor/new_join.go (hunk 13 above)
Comment (hanfei1991) [inline executor/new_join.go]: outerCursor should increase one by one to construct rows with null.

File: executor/new_join.go [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+	e[i], e[j] = e[j], e[i]
+}
+
+// Less implements sort.Interface Less interface.
+func (e orderedRows) Less(i, j int) bool {
+	return bytes.Compare(e[i].key, e[j].key) < 0
+}
+
+// IndexLookUpJoin fetches batches of data from outer executor and constructs ranges for inner executor.
+type IndexLookUpJoin struct {
```
Comment (XuHuaiyu) [inline executor/new_join.go]: rename it to IndexLoopUpJoinExec ?

Comment (hanfei1991) [inline executor/new_join.go]: I think the exec suffix is unnecessary.  The type is executor and there are no name conflicts with it

File: executor/new_join.go [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+					datum, _ := col.Eval(outerRow.Data)
+					joinDatums = append(joinDatums, datum)
+				}
+				joinOuterEncodeKey, err := codec.EncodeValue(nil, joinDatums...)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				e.outerRows = append(e.outerRows, orderedRow{key: joinOuterEncodeKey, row: outerRow})
+				e.innerDatums = append(e.innerDatums, orderedRow{key: joinOuterEncodeKey, row: &Row{Data: joinDatums}})
+			} else if e.outer {
```
Comment (XuHuaiyu) [inline executor/new_join.go]: If `match` is false, why should we still remain the outerRow when it is an OuterJoin.

Comment (hanfei1991) [inline executor/new_join.go]: You mean we should return it?

Comment (XuHuaiyu) [inline executor/new_join.go]: I mean continue to get the next OuterRow which matches the leftConditions.

Comment (hanfei1991) [inline executor/new_join.go]: Outerjoin should return every outer row

Comment (XuHuaiyu) [inline executor/new_join.go]: I know what you mean here, 
but if it doesn't match the `leftConditions`, we still return it?

Comment (hanfei1991) [inline executor/new_join.go]: Left conds are part of on conditions

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**